### PR TITLE
V1 and Count

### PR DESCRIPTION
**Pull request by davidhadas**
# Security Data Package

This PR includes V1 the common interface to all data types managed by Guard. 

The v1.go includes three interfaces:

- Profile: describing a sample of a data type
- Pile: accumulating multiple samples of a data type to enable learning.
- Config: the rules describing what is expected from the data type.

A first data type count.go is included as a reference to show how a simple uint8 counter is managed.

### CODE REVIEW

**Comment by davidhadas** [comment]
/hold
An upcoming update to this PR with improved v1_tests is in the works

**Review by psschwei — COMMENTED** [review]
Here's part 1 of my review. Will finish up with v1_test.go tomorrow.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:15]: It is used as MyData.Profile(<Here I put my data that is being profiled>)

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:40]: depth represents how many "tabs" we ned to shift to print the data
here it is not used, but with more complex data types that print multiple lines (think of lists or key-val objects), we will use depth

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:24]: xxxPile Data types will pile the dataType, but it will not always be as simple as an array
We should really keep each of the data types 
- xxxProfile
- xxxPile
- xxxConfig   
as a standalone and not derived from another

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:69]: That will indeed be more aligned with go standard  
However, it will make it harder to accumulate multiple issues found in a tree of Data Types. 

I think we need to consider this in the next PR when we see a Data Type that relies on count (depth of 2), 
Ideally, we will define a mechanism that will allow us to accumulate issues (not great with errors, very poor but can be used on day 1 with strings)

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:27]: The function signatures are all defined in v1.go
Same signatures will be used for other data types

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:135]: same as above - count is an example of a simple data type that does not use depth
The function signatures are shared between all data types

File: pkg/apis/wsecurity/v1/README.md
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/README.md:3]: Do you want to link to a design or architecture document?

File: pkg/apis/wsecurity/v1/README.md
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/README.md:25]: Is the intent that the Pile is _empty_ or _unusable_?

If the latter, is there a particular reason to prevent the reuse of `Pile`s?

File: pkg/apis/wsecurity/v1/README.md
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/README.md:26]: Does this consume the `Pile`, or no? Given the previous `Merge` operation, it seems worth specifying.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go:13]: In particular, this code will `panic` if the `args[0]` is not a `uint8`, and ignore subsequent arguments.

The interface here feels like it's actively fighting against Go's type system to make it something more like Javascript. I'm not sure I understand what the value of this `Profile` interface is.

File: pkg/apis/wsecurity/v1/v1.go
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/v1.go:13]: Use generics here?

File: pkg/apis/wsecurity/v1/v1.go
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/v1.go:9]: The `depth` argument for `String()` is surprising, since a `String` method without a depth argument will cause the object to automatically meet the `fmt.Stringer` interface, which is pretty useful.

Additionally, we tend to prefer structured (JSON format) logging from components via zap.logger, and it feels like multi-line indentation would be less useful given structured logs.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go:24]: Do you want to use Go 1.18 generics (rather than using concrete types here?

e.g.
```golang
type Profile[T any] struct {
  T value
}

func (p *Profile[T]) Profile(args ...interface{}) {
  // ...
}

// OR

func (p *Profile[T]) Profile(args ...T) {
  // ...
}
```

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go:27]: Again, I think using generics will simplify this, and also make it harder to accidentally `CountPile.Add(StringProfile)`, which looks like it will explode with a `panic` at the moment.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/count.go:37]: The documentation claims that `*otherPile` will be unusable after this operation, but it looks like it will be usable but _may alias some values with `pile`_.

Why not call `otherPile.Clear()` at the end of this?

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go:69]: Take a look at the pattern here for errors: https://github.com/knative/pkg/blob/main/apis/field_error.go#L63

In particular, you can define an accumulator for errors like so: https://github.com/knative/pkg/blob/main/apis/field_error.go#L196

```golang
func (fe *FieldError) Also(errs ...*FieldError) *FieldError {
  // Note that `fe` may be `nil` here
```

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/count.go:78]: I'd extract the profile value prior to the loop, and then use the `for _, item := range(*config)` format, so your inner comparison looks like:

```golang
  if profile < item.Min {
    break
  }
  if profile <= item.Max {
    return ""
  }
```

If you're particularly concerned about efficiency, you could actually do a binary search or heuristic search across the values if you coalesce and sort in `Learn`. Right now, the set of ranges will just keep growing from each `Learn`, so you can probably only accumulate for a certain amount of time before this slows down.

File: pkg/apis/wsecurity/v1/v1_test.go
Comment (evankanderson) [inline pkg/apis/wsecurity/v1/v1_test.go:50]: I don't understand how these permutations were arrived at. Also, why call `.String(9)` without testing the output? I'd expect `String` to not change the value; if these are simply testing for non-panic, it seems like a more focused test would make more sense.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:24]: This is an interesting discussion to have.

When can generics help?
Generics can help when we have multiple data types that can share the **same** implementation. 
SG uses multiple data types (more than 14 right now) that each requires its own very different implementation... but they are all sharing the same basic interface.

E.g. `XXXConfig.Decide()` if very different in its implementation from `YYYConfig.Decide()`. The same is true for other methods of the XXXConfig/Pile/Profile and YYYConfig/Pile/Profile data types.

So using generics may not offer us any immediate code reduction, but can help make some of the more basic data types (such as Count here) more generic :). E.g. Count can be made to support uint8, unit16, and uint32 counters in one stroke. 

For example:
```golang
type CountPile[T any] []T

func (pile *CountPile[T]) Merge(otherValPile ValuePile) {
	otherPile := otherValPile.(*CountPile[T])
	*pile = append(*pile, *otherPile...)
}

var a,b CountPile[uint8]
...
a.Merge(&b)
```

---
SG uses Kubernetes gengo, right now it does not seem to support generics code. 
We can add this as we move forward.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:13]: This is an unfortunate truth...

The previous version of SG does not use a central interface for the dataTypes. Instead, there are some 15 data types, which all were designed with very similar methods (such as Decide, Learn, Merge, Add etc.). Each dataType uses a different method signature - for example:
```golang
func (p *KeyValPile) Add(kv *KeyValProfile) {...}
func (p *SimpleValPile) Add(svp *SimpleValProfile) {...}
func (p *UrlPile) Add(u *UrlProfile) {...}
```

I refactored the code in this version to use a central interface   - i.e. `v1.go`

This aligns all data types to a single interface which has many advantages in my mind. The obvious disadvantage is what you bring up here - it moves the data type assertion from compile time to run time. I guess we can decide to avoid a central interface and go with the code as is from the previous version - but I am not sure this is a better path.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:69]: So, error accumulation may be another option. 

Let's delay this discussion to the next PR though where we get to see 2 levels of data types.

We need to discuss if to accumulate and what is the best way.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:27]: As discussed above, each method is implemented differently on each data type,  in this case, do you still see a way for generics to help remove the runtime type assertions?

It seems that even when we use generics, we will still need to have different type assertions for each implementation, simply because each of them does something different with the data... so:

```golang
func (p *KeyValPile) Add(...) // need to assert that the data is of type KeyValProfile}
func (p *ReqPile) Add(...) // need to assert that the data is of type ReqProfile
```



See above - I can't see how generics help us in simplifying or removing any code.

File: pkg/apis/wsecurity/v1/README.md
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md:3]: going forward, yes.

File: pkg/apis/wsecurity/v1/README.md
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md:25]: The intent of the `no longer usable` is to free us to use internal data constructs of the other Pile if and when we choose to do so implementation-wise (to reduce overhead). 

The expected behavior is for piles to accumulate in QP instances  - then be delivered to a remote central learner service which will merge Piles from different  QP instances. once they are merged, these piles are dumped anyhow. I.e. the code looks like that:
```golang
... 
// get a pile sent from a QP instance
err := json.NewDecoder(req.Body).Decode(&pile)
...
// add the pile to a bigger pile we keep for all instances that serve the Knative service XYZ
record.pile.Merge(&pile) // pile no longer used
...
```

File: pkg/apis/wsecurity/v1/README.md
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md:26]: When I wrote 'consume' I meant to say that some of the data in someOtherPile may now be reused by pile, to avoid deep copy when possible.

File: pkg/apis/wsecurity/v1/v1.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/v1.go:13]: See above

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:37]: `but it looks like it will be usable but may alias some values with pile` - correct, but we should no longer use pile without creating new values in pile  - we should not change one pile via changing the other.

Think of the dataTypes as offering many layers where each layer is implemented by either basic data types or other layers. It is therefore enough to call Clear() on the top layer, rather than calling it in every piece of data in each layer. 

Calling `otherPile.Clear()` at the end of any implementation of Merge() will result in many redundant calls for `Clear()`. It may be better to leave this to the user of this package - the user may call `Clear()` once on the top level dataType after calling the top level `Merge()` or decide not to call it if it will no longer be used.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:78]: We should leave optimizations such as binary search to a later stage, I am not sure we will need it specifically with uint8 ranges. The typical number of ranges I would expect is 1 or very few. If that is not the case we need to fix Learn() to generalize the data - which is planned anyhow for the next version of SG, but let's get this version up first.

File: pkg/apis/wsecurity/v1/v1.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/v1.go:9]: Yep, this is probably a not-so-great choice made earlier in the design and I never got to rethink it.

Let's define the problem at hand:
GuardianConfig includes a ReqConfig which includes a list of HeadersConfig, each header has a key and a value, each value has a list of counters and each counter has a list of ranges (MinMax). 

It is very useful to be able to print a GuradianConfig (and other structures like it) to log during debug - so we need some way to implement `GuradianConfig.stringify()`. 

Each dataType needs to create its own part, and all together need to create a string that we can log (best if we can log it nicely both in production and in development with multi-line) - we are actually building a structure somewhat like a JSON  - so maybe we simply need to build a JSON. 

The depth used today is because we are building it to print more as a JSON prettify would do. Maybe better if we build a string that is a legal JSON and use some lib to prettify it when needed.

File: pkg/apis/wsecurity/v1/v1_test.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/v1_test.go:50]: We need more focused tests. 

Maybe it is best to leave this as is for now and address it in a separate PR. I think there is more to be done in improving the current tests.

(We do have many focused tests already, these were added to capture some additional code branches like when you Sting() a freshly created object )

**Review by davidhadas — COMMENTED** [review]
Move to Marshal?

File: pkg/apis/wsecurity/v1/count.go
Comment (psschwei) [reply pkg/apis/wsecurity/v1/count.go:69]: My only worry in waiting is if we change the output of an exported v1 API, does that require a new v2 version?  I know it'll likely have little impact right now (so I may be over thinking this), but technically v1 does require backwards compatibility so would this be considered a "breaking" change...

File: pkg/apis/wsecurity/v1/count.go
Comment (psschwei) [reply pkg/apis/wsecurity/v1/count.go:24]: > SG uses multiple data types (more than 14 right now) that each requires its own very different implementation

Just thinking out loud, but would it be possible to come up with a more generic data model that can better accommodate those different existing types? (Good chance there's not, but this might be the easiest time to revisit the data model if there is, so figured it was worth asking)

Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:24]: We can revisit this question at the next PR when we get to see 3 more data types: 
- SimpleVal which uses Count (shown here), AsciiFlags, and FlagsSlice

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go:69]: We are long before our Guard v0.1
As for API version, I did not answer the above since I am not sure how to relate the API version to the Guard version. 
I think we will have many Guard versions and potentially more than one API version - and I am not sure what the correct practice is as for API version - will try to read about that and check if we can move to a v0.1 API version

File: pkg/apis/wsecurity/v1/README.md
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md:9]: I will have a look at what we should be using as a first API version that is alpha

Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md:9]: We can go with v1alpha1

File: pkg/apis/wsecurity/v1/count_test.go
Comment (psschwei) [inline pkg/apis/wsecurity/v1/count_test.go:29]: nit: we're only able to run the whole suite of tests this way... for example, `go test -run Basics ./...` will come back with "no tests to run". That's probably fine for now, but may be something we want to revisit down the line.

File: pkg/apis/wsecurity/v1/count.go
Comment (psschwei) [reply pkg/apis/wsecurity/v1/count.go:69]: in that case, should we change the path so that it's not `.../v1/...`? I think that was the main thing that was hanging me up

File: pkg/apis/wsecurity/v1/count_test.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count_test.go:29]: In a way, the v1_tests are the basic tests, while we may need additional data type-specific tests to cover code path which is specific to the data-type. It is true that one of the benefits of going to an interface-based set of data types is that it reduces the need to write testing per data type.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go]: Does it make sense to include these to make it more clear what varies?

The simplicity of the existing operations makes it hard to understand the end-goal state.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go]: I think this is where type constraints with generics could be helpful -- You might be able to declare an interface like:

```golang
type Pile[T any] interface {
  Add(Profile[T])
  Clear()
  fmt.Stringer
}
```

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go]: I'd remove the version number for the v0.x version. You can add the version numbering and enjoy `go mod` versioning once you have a breaking change _after_ v1.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go]: If `KeyValPile` implements `Pile[KeyVal]`, and `ReqPile` implements `Pile[Req]`, then they would have different arguments, but implement the same interface.

File: pkg/apis/wsecurity/v1/README.md
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/README.md]: "Learn a new config rules based on a pile" doesn't say whether it consumes the pile or not.

File: pkg/apis/wsecurity/v1/v1.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/v1.go]: I'd implement [the `json.Marshaller` interface](https://pkg.go.dev/encoding/json#Marshaler)

Comment (evankanderson) [reply pkg/apis/wsecurity/v1/v1.go]: Note that you also may want to implement `fmt.Stringer`, even if it just calls `MarshalJSON` under the covers.

File: pkg/apis/wsecurity/v1/count.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count.go]: The cost of calling `Clear` appears to be a single pointer assignment. The cost of accidentally re-using a `Pile` is silent data corruption.

If we _are_ concerned about performance here, I can go over the design with an eye to performance, and suggest some benchmarks to write. There are a number of cases here where it seemed like we weren't hugely performance-sensitive (e.g. not sorting the ranges in CountPiles).

File: pkg/apis/wsecurity/v1/count_test.go
Comment (evankanderson) [reply pkg/apis/wsecurity/v1/count_test.go]: I'd be +1 on removing these tests until we've firmed up the interface (possibly by introducing the additional data types and accumulators mentioned in one of the above comments in this PR).

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: It seems as indicated above that generics will not change the number of core dataTypes we need to support since their implementation is quite different from one another, learning an HttpHeader is very different from learning a Counter or learning a URL. 
Generics will make each core dataType become more generic which may become useful at some point.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: @evankanderson  - what generics let you do is have the same Add function (same lines of code) implemented once for all dataTypes. Since we need different Add functions (different lines of code for each function, regardless of the type being added) implementing one generic `func Pile[T] Add()` function will not help us.

File: pkg/apis/wsecurity/v1/README.md
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/README.md]: (tnx, I originally thought the language and using 'consume'  was unclear..   will be fixed)

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: Agreed.  

Rough estimate - we are going to have some 1000 counters per session (Req-Resp)
Each counter will typically have zero or one or two ranges - no more. 
Yes, specifically on these widely used Counters we need to be extra careful with lines of code executed.

I think we can either:
- decide to DeepCopy always and not affect other items
- decide to Clear other items always
- decide to Clear other items only if we consume them such that we never reach silent data corruption
- decide to make only some DataTypes exposed  (e.g. avoid exposing low-level data types and do Clear() only on exposed DataTypes
- mark the other data type as "consumed" (not sure if this will make more sense than simply clearing them) 
 
also - we need to add a list of "Things to revisit" into the repo such that we do not lose sight of those.

File: pkg/apis/wsecurity/v1/count_test.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count_test.go]: I'd prefer to have a list of items to revisit since I am starting from a full repository and cutting slices in, adjusting all codes to the standards we mutually agree on along the way.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: sure, expect a new PR soon with these as well later today

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: We moved to v1aplha1 as customary in Kube.

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: Added to TODO list

File: pkg/apis/wsecurity/v1/count.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1/count.go]: Generics added to TODO under Future

**Comment by davidhadas** [comment]
Seems like we cn merge this one and continue - all broader items were added to a TODO file
/unhold

**Review by maximilien — APPROVED** [review]
/approve
/lgtm

**Comment by evankanderson** [comment]
Looking at 16, it seems like we have the following setup:

* `XXProfile` is a concrete type which interprets `Profile()` specially for each type.
* `XXPile` is a "collection" of `XXProfile` objects. In some cases, there is an optimized mechanism for merging the profiles (equivalent to a "set", but often using bitmaps); the default is to simply collect them in a slice.  In particular, I can imagine a future algorithm which is concerned about the _frequency_ of profiles in a pile, so optimizing this to a "set" is probably not a good idea -- a histogram might be acceptable, but "collection" seems about the right scope.
* `XXConfig` performs profile-specific operations for `Decide` and `Learn` as well as `Fuse` (though I don't see why `Fuse` is needed at the moment).

I wonder whether making `Pile` a generic interface which collects in a slice with an _optional_ `Coalesce` method on `XXProfile` would work:

https://go.dev/play/p/bpzlStlyhB5

**Comment by davidhadas** [comment]
> Looking at 16, it seems like we have the following setup:
> 
> * `XXProfile` is a concrete type which interprets `Profile()` specially for each type.
> * `XXPile` is a "collection" of `XXProfile` objects. In some cases, there is an optimized mechanism for merging the profiles (equivalent to a "set", but often using bitmaps); the default is to simply collect them in a slice.  In particular, I can imagine a future algorithm which is concerned about the _frequency_ of profiles in a pile, so optimizing this to a "set" is probably not a good idea -- a histogram might be acceptable, but "collection" seems about the right scope.
> * `XXConfig` performs profile-specific operations for `Decide` and `Learn` as well as `Fuse` (though I don't see why `Fuse` is needed at the moment).
> 
> I wonder whether making `Pile` a generic interface which collects in a slice with an _optional_ `Coalesce` method on `XXProfile` would work:
> 
> https://go.dev/play/p/bpzlStlyhB5

I am moving this to PR/16 as we merged PR/9

**Pull request merged**
